flocker/control/test/test_persistence.py

```diff
@@ -363,6 +367,52 @@ def saved(_):
         d.addCallback(saved)
         return d
 
+    @validate_logging(assertHasMessage, _LOG_UNCHANGED_DEPLOYMENT_NOT_SAVED)
+    def test_callback_not_called_for_unchanged_deployment(self, logger):
+        """
+        If the old deployment and the new deployment are equivalent, registered
+        callbacks are not called.
+        """
+        service = self.service(FilePath(self.mktemp()), logger)
+
+        state = []
+
+        def callback():
+            state.append(None)
+
+        saving = service.save(TEST_DEPLOYMENT)
+
+        def saved_old(ignored):
+            service.register(callback)
+            return service.save(TEST_DEPLOYMENT)
+
+        saving.addCallback(saved_old)
+
+        def saved_new(ignored):
+            self.assertEqual(
+                [], state,
+                "Registered callback was called; should not have been."
+            )
+
+        saving.addCallback(saved_new)
+        return saving
+
+    def test_success_returned_for_unchanged_deployment(self):
+        """
+        ``save`` returns a ``Deferred`` that fires with ``None`` when called
+        with a deployment that is the same as the already-saved deployment.
+        """
+        service = self.service(FilePath(self.mktemp()), None)
+
+        saving = service.save(TEST_DEPLOYMENT)
+
+        def saved_old(ignored):
+            saving = service.save(TEST_DEPLOYMENT)
+            saving.addCallback(self.assertEqual, None)
+
```

Comment:
Need to return `saving` here. And perhaps give it different name so this is less confusing.